### PR TITLE
[DO NOT MERGE] ci: optimize and unify scripts and build process

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# What ❔

* [x] Unify and update CI scripts: introduce matrix jobs, use common integration tests and benchmarks (remove necessity to create special configs)
* [x] Rename `ZKEVM` to `ZKsync` everywhere
* [x] Fix versioning - use only the second part of the tag as the `ZKsync` version
* [x] Make the ZKsync part of the changelog more unified

## Why ❔

Fixes [CPR-1689](https://linear.app/matterlabs/issue/CPR-1689/update-ci-of-era-solidity-repository)

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.

### CODE REVIEW

File: liblangutil/Token.h
Comment (antonbaliasnikov) [inline liblangutil/Token.h]: `uint8_t`, etc. types are not available without this header on `mingw` platform. It is required for Windows compilation.

File: libyul/CMakeLists.txt
Comment (antonbaliasnikov) [inline libyul/CMakeLists.txt:67]: I've put `EraVM` here instead of `ZKsync`. I believe it's the right place to use VM type?

Comment (hedgar2017) [reply libyul/CMakeLists.txt:67]: Yes, awesome!

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Ready after the changelog tweak!

**Comment by antonbaliasnikov** [comment]
> Ready after the changelog tweak!

This is done.

**Review by hedgar2017 — DISMISSED** [review]
Hey @abinavpp, we've moved the changelog update to this branch in order to minimize your patch.

@antonbaliasnikov please be careful with the global substitution of versions!

**Review by hedgar2017 — APPROVED** [review]

File: .github/workflows/build-release.yaml
Comment (abinavpp) [inline .github/workflows/build-release.yaml]: Are we building < 0.6.0 binaries for Windows? The original build_and_release_binary didn't.

File: .github/workflows/build-release.yaml
Comment (abinavpp) [inline .github/workflows/build-release.yaml]: I recall build errors with newer gcc (like the default one in Ubuntu-22.04) when building older versions. IIRC, gcc-9.4 was able to build older versions. If we're using this configuration for all versions, can you confirm if we're able to build all of them? Thanks!

File: .github/actions/build/action.yml
Comment (abinavpp) [inline .github/actions/build/action.yml]: I see the "10" in the original as well. Is it possible to not limit the job count here? (I'm not sure if just "-j" in cmake --build is same as just "-j" in make).

File: .github/workflows/build-release.yaml
Comment (hedgar2017) [reply .github/workflows/build-release.yaml]: Indeed, @antonbaliasnikov we do not support <0.6.0 on Windows. Devops said it was too problematic.

File: .github/actions/build/action.yml
Comment (hedgar2017) [reply .github/actions/build/action.yml]: True, let's not hardcode it and look up the thread count.

File: .github/workflows/build-release.yaml
Comment (hedgar2017) [reply .github/workflows/build-release.yaml]: @abinavpp though probably Anton wants to take it into account in each particular branch.
Let's wait for his response.

Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: > Devops said it was too problematic.

I completely reworked Windows build, no crazy powershell anymore, we do mingw64 and clang as for zksolc/zkvyper. So, in theory, there should be no difference, but I'll test it properly today and check if there are issues with older versions.

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: Regardless of the newer Docker image, we still set `gcc-9` as the default one:
https://github.com/matter-labs/compiler-infra/blob/main/images/llvm_runner/ubuntu22-llvm17.Dockerfile#L58-L61

so, there should be no difference. But I'll test it thoroughly before merging.

File: .github/actions/build/action.yml
Comment (antonbaliasnikov) [reply .github/actions/build/action.yml]: This one is tricky... even `make -j` is not working by default on all platforms and requires the number of threads to be explicitly specified. At the same time, defining the number of CPU threads for all platforms differs. I'll think about something universal or introduce a new platform-specific step.

Comment (hedgar2017) [reply .github/actions/build/action.yml]: Let's do properly with a platform specific step. On Linux and MacOS it's easy, and for Windows... hardcode it deserves :)

**Review by hedgar2017 — DISMISSED** [review]

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: After lots of testing and investigating the old builds, I made a decision to return the ugly powershell, because it's impossible to make mingw64 work with C++ code which is 5 years old and build system barely working :( 

Check for 0.6.0 added back.

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: I've introduced a new public docker image `matterlabs/llvm_runner:ubuntu20-llvm17-latest` specially for this case.

File: .github/actions/build/action.yml
Comment (antonbaliasnikov) [reply .github/actions/build/action.yml]: Completed.

**Pull request closed without merging**
